File: ScaleformUI_Lua/src/Menus/PauseMenus/PauseMenu/Tabs/SubmenuTab.lua
SubmenuTab = {}
SubmenuTab.__index = SubmenuTab
setmetatable(SubmenuTab, { __index = BaseTab })
SubmenuTab.__call = function() return "SubmenuTab" end

---@class SubmenuTab
function SubmenuTab.New(name, color)
    local data = BaseTab.New(name, color)
    data._identifier = "Page_Info"
    data.LeftColumn = SubmenuLeftColumn.New(0)
    data.CenterColumn = SubmenuCentralColumn.New(1)
    local meta = setmetatable(data, SubmenuTab)
    meta.LeftColumn.Parent = meta
    meta.CenterColumn.Parent = meta
    return meta
end

function SubmenuTab:currentItemType()
    return self.LeftColumn:currentItemType()
end

function SubmenuTab:SwitchColumn(index)
    local col = self:GetColumnAtPosition(index)
    self.CurrentColumnIndex = index
    if self.Parent ~= nil and self.Parent:Visible() then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_MENU_LEVEL", self.CurrentColumnIndex + 1)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("MENU_SHIFT_DEPTH", 0, true, true)
        self.Parent.focusLevel = self.CurrentColumnIndex + 1
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_HIGHLIGHT", col.position, col:Index(), true, true)
        col.Items[col:Index()]:Selected(true)
    end
end

function SubmenuTab:AddLeftItem(item)
    item.ParentTab = self
    self.LeftColumn:AddItem(item)
end

function SubmenuTab:StateChange(state)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("MENU_STATE", self:currentItemType())
    self.CenterColumn:Clear()
    if state ~= 0 then
        for k,v in ipairs(self.LeftColumn.Items[self.LeftColumn:Index()].ItemList) do
            self.CenterColumn:AddItem(v)
        end
    end
    for k,v in pairs(self.CenterColumn.Items) do
        v.ParentColumn = self.CenterColumn
    end
    if self:currentItemType() == LeftItemType.Statistics then
        self.CenterColumn.VisibleItems = 16
        self.CenterColumn:InitColumnScroll(true, 2, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER)
        self.CenterColumn:SetColumnScroll(-1,-1,-1,"", #self.CenterColumn.Items<self.CenterColumn.VisibleItems)
    elseif self:currentItemType() == LeftItemType.Settings then
        self.CenterColumn.VisibleItems = 16
        self.CenterColumn:InitColumnScroll(true, 2, ScrollType.ALL, ScrollArrowsPosition.CENTER)
        self.CenterColumn:SetColumnScroll(self.CenterColumn:Index(),#self.CenterColumn.Items,self.CenterColumn.VisibleItems,"", #self.CenterColumn.Items<self.CenterColumn.VisibleItems)
    elseif self:currentItemType() == LeftItemType.Info then
        self.CenterColumn.VisibleItems = 10
        self.CenterColumn:InitColumnScroll(true, 2, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER)
        self.CenterColumn:SetColumnScroll(-1,-1,-1,"", #self.CenterColumn.Items<self.CenterColumn.VisibleItems)
    elseif self:currentItemType() == LeftItemType.Keymap then
        self.CenterColumn.VisibleItems = 15
        self.CenterColumn:InitColumnScroll(true, 2, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER)
        self.CenterColumn:SetColumnScroll(-1,-1,-1,"", #self.CenterColumn.Items<self.CenterColumn.VisibleItems)
    else
        self.CenterColumn.VisibleItems = 0
    end
end

function SubmenuTab:GoUp()
    if not self.Focused then return end
    if self.CurrentColumnIndex == 0 then
        self.LeftColumn:GoUp()
        self.CenterColumn:currentColumnType(self:currentItemType())
        self:StateChange()
        self:Refresh(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_INPUT_EVENT", self.CurrentColumnIndex, 8)
    elseif self.CurrentColumnIndex == 1 then
        self.CenterColumn:GoUp()
        if self.CenterColumn:currentColumnType() == LeftItemType.Settings then
            self.CenterColumn:SetColumnScroll(self.CenterColumn:Index(),#self.CenterColumn.Items,self.CenterColumn.VisibleItems,"", #self.CenterColumn.Items<self.CenterColumn.VisibleItems)
        end
    end
end

function SubmenuTab:GoDown()
    if not self.Focused then return end
    if self.CurrentColumnIndex == 0 then
        self.LeftColumn:GoDown()
        self.CenterColumn:currentColumnType(self:currentItemType())
        self:StateChange()
        self:Refresh(false)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_INPUT_EVENT", self.CurrentColumnIndex, 9)
    elseif self.CurrentColumnIndex == 1 then
        self.CenterColumn:GoDown()
        if self.CenterColumn:currentColumnType() == LeftItemType.Settings then
            self.CenterColumn:SetColumnScroll(self.CenterColumn:Index(),#self.CenterColumn.Items,self.CenterColumn.VisibleItems,"", #self.CenterColumn.Items<self.CenterColumn.VisibleItems)
        end
    end
end

function SubmenuTab:GoLeft()
    if not self.Focused then return end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_INPUT_EVENT", self.CurrentColumnIndex, 10)
    if self.CurrentColumnIndex == 1 then
        self.CenterColumn:GoLeft()
    end
end

function SubmenuTab:GoRight()
    if not self.Focused then return end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_INPUT_EVENT", self.CurrentColumnIndex, 11)
    if self.CurrentColumnIndex == 1 then
        self.CenterColumn:GoRight()
    end
end

function SubmenuTab:Select()
    if not self.Focused then return end
    if self.CurrentColumnIndex == 0 then
        if self:currentItemType() ~= LeftItemType.Settings then return end
        local lI = self.LeftColumn.Items[self.LeftColumn:Index()]
        if not lI:Enabled() then
            PlaySoundFrontend(-1, "ERROR","HUD_FRONTEND_DEFAULT_SOUNDSET", false)
            return
        end
        self.CurrentColumnIndex = self.CurrentColumnIndex + 1
        local ret = true
        local idx = 1
        for k,v in pairs(lI.ItemList) do
            if v.Enabled then
                ret = false
                idx = k
                break
            end
        end
        if ret then return end
        self.CenterColumn:Index(idx)
        self.Parent:FocusLevel(self.Parent.focusLevel + 1)
        self.Parent.OnColumnItemSelect(self.Parent, self, self.LeftColumn, self.LeftColumn:Index())
    elseif self.CurrentColumnIndex == 1 then
        self.CenterColumn:Select()
        self.Parent.OnColumnItemSelect(self.Parent, self, self.CenterColumn, self.CenterColumn:Index())
    end
end

function SubmenuTab:GoBack()
    if not self.Focused then return end
    if self.CurrentColumnIndex == 1 then
        self.CurrentColumnIndex = self.CurrentColumnIndex - 1
        self.Parent:FocusLevel(self.Parent.focusLevel - 1)
    end
end

function SubmenuTab:MouseEvent(eventType, context, index)
    if not self.Focused then return end
    if eventType == 5 then
        if self.CurrentColumnIndex == context then
            if self:CurrentColumn():Index() ~= index then
                if self.CurrentColumnIndex == 0 then
                    self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(false)
                    self.LeftColumn:Index(index)
                    self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(true)
                    self:StateChange(self:currentItemType())
                    self:Refresh(false)
                elseif self.CurrentColumnIndex == 1 then
                    self.CenterColumn.Items[self.CenterColumn:Index()]:Selected(false)
                    self.CenterColumn:Index(index)
                    self.CenterColumn.Items[self.CenterColumn:Index()]:Selected(true)
                end
                return
            end
            self:Select()
        else
            if context > self.CurrentColumnIndex then
                self.Parent:FocusLevel(self.Parent:FocusLevel() + 1)
                self.CurrentColumnIndex = self.CurrentColumnIndex + 1
            elseif context < self.CurrentColumnIndex then
                self.Parent:FocusLevel(self.Parent:FocusLevel() - 1)
                self.CurrentColumnIndex = self.CurrentColumnIndex - 1
            end
            if self.CurrentColumnIndex == 0 then
                self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(false)
                self.LeftColumn:Index(index)
                self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(true)
                self:StateChange(self:currentItemType())
                self:Refresh(false)
            elseif self.CurrentColumnIndex == 1 then
                self.CenterColumn.Items[self.CenterColumn:Index()]:Selected(false)
                self.CenterColumn:Index(index)
                self.CenterColumn.Items[self.CenterColumn:Index()]:Selected(true)
            end
        end
    elseif eventType == 10 or eventType == 11 then
        local dir = -1
        if eventType == 11 then
            dir = 1
        end
        self:MouseScroll(dir)
    end
end

function SubmenuTab:MouseScroll(dir)
    local col = self.Parent.hoveredColumn
    if self.CurrentColumnIndex == 0 then
        if col == 1 then
            if self:currentItemType() == LeftItemType.Info or self:currentItemType() == LeftItemType.Statistics then
                PlaySoundFrontend(-1, "NAV_UP_DOWN","HUD_FRONTEND_DEFAULT_SOUNDSET", false)
                return
            end
        end
        if dir == -1 then
            self.LeftColumn:GoUp()
        else
            self.LeftColumn:GoDown()
        end
        self.CenterColumn:currentColumnType(self:currentItemType())
        self:StateChange(self:currentItemType())
        self:Refresh(false)
    elseif self.CurrentColumnIndex == 1 then
        if self:currentItemType() == LeftItemType.Settings then
            if dir == -1 then
                self.CenterColumn:GoUp()
            else
                self.CenterColumn:GoDown()
            end
            if self.CenterColumn:currentColumnType() == LeftItemType.Settings then
                self.CenterColumn:SetColumnScroll(self.CenterColumn:Index(),#self.CenterColumn.Items,self.CenterColumn.VisibleItems,"", #self.CenterColumn.Items<self.CenterColumn.VisibleItems)
            end
            PlaySoundFrontend(-1, "NAV_UP_DOWN","HUD_FRONTEND_DEFAULT_SOUNDSET", false)
        end
    end
end

function SubmenuTab:Focus()
    BaseTab.Focus(self)
    self.LeftColumn:Index(self.LeftColumn.index)
    self.LeftColumn:HighlightColumn(true, false, true)
    self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(true)
    self:Refresh(true)
end

function SubmenuTab:UnFocus()
    if self.CurrentColumnIndex > 0 then
        self.Parent:FocusLevel(self.Parent:FocusLevel() - 1)
        self.CurrentColumnIndex = self.CurrentColumnIndex - 1
    end
    self.LeftColumn.Items[self.LeftColumn:Index()]:Selected(false)
    BaseTab.UnFocus(self)
end

function SubmenuTab:Refresh(highlightOldIndex)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("ALLOW_CLICK_FROM_COLUMN", 0, true)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_DATA_SLOT_EMPTY", 1)
    self.CenterColumn:Index(1)
    for i=1, #self.CenterColumn.Items, 1 do
        self:SetDataSlot(1, i)
    end
    if self:currentItemType() == LeftItemType.Keymap then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_TITLE", 1, self.LeftColumn.Items[self.LeftColumn:Index()].TextTitle, self.LeftColumn.Items[self.LeftColumn:Index()].KeymapRightLabel_1, self.LeftColumn.Items[self.LeftColumn:Index()].KeymapRightLabel_2)
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_FOCUS", 1, false, false, false)
    elseif self:currentItemType() == LeftItemType.Settings then
        ScaleformUI.Scaleforms._pauseMenu._pause("SET_COLUMN_HIGHLIGHT", 1, self.CenterColumn:Index() - 1, true, true)
    end
    self.CenterColumn:ShowColumn()
end

function SubmenuTab:Populate()
    local item = self.LeftColumn.Items[self.LeftColumn:Index()]
    item:Selected(true)
    self.CenterColumn:Clear()
    if self:currentItemType() ~= LeftItemType.Empty then
        for k,v in ipairs(self.LeftColumn.Items[self.LeftColumn:Index()].ItemList) do
            v.ParentColumn = self.CenterColumn
            self.CenterColumn:AddItem(v)
        end
    end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_MENU_LEVEL", 1)
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("MENU_STATE", self:currentItemType())
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_MENU_LEVEL", 0)
    for i=1, #self.LeftColumn.Items, 1 do
        self:SetDataSlot(0, i)
    end
    for i=1, #self.CenterColumn.Items, 1 do
        self:SetDataSlot(1, i)
    end
end

function SubmenuTab:ShowColumns()
    self.LeftColumn:ShowColumn()
    self.CenterColumn:ShowColumn()
    if self:currentItemType() == LeftItemType.Settings then
        ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_STATE", 0)
    end
    ScaleformUI.Scaleforms._pauseMenu._pause:CallFunction("SET_COLUMN_FOCUS", 0, false, false, false)
    self.LeftColumn:InitColumnScroll(true, 1, ScrollType.UP_DOWN, ScrollArrowsPosition.RIGHT)
    self.LeftColumn:SetColumnScroll(self.LeftColumn:Index(), #self.LeftColumn.Items, 16, "", #self.LeftColumn.Items < 16)
end

function SubmenuTab:SetDataSlot(slot, index)
    if slot == 0 then
        self.LeftColumn:SetDataSlot(index)
    elseif slot == 1 then
        self.CenterColumn:SetDataSlot(index)
    end
end

function SubmenuTab:UpdateSlot(slot, index)
    if slot == 0 then
        self.LeftColumn:UpdateSlot(index)
    elseif slot == 1 then
        self.CenterColumn:UpdateSlot(index)
    end
end

File: ScaleformUI_Csharp/Menus/Pause Menus/PauseMenu/Tabs/SubmenuTab.cs
using ScaleformUI.Elements;
using ScaleformUI.Menu;
using ScaleformUI.Menus;
using ScaleformUI.PauseMenus.Elements;
using ScaleformUI.PauseMenus.Elements.Columns;
using System.Linq;
using System.Reflection.Emit;
using static CitizenFX.Core.UI.Screen;

namespace ScaleformUI.PauseMenu
{
    public class SubmenuTab : BaseTab
    {
        internal LeftItemType currentItemType => ((SubmenuLeftColumn)LeftColumn).currentItemType;

        public SubmenuTab(string name, SColor color) : base(name, color)
        {
            _type = 1;
            _identifier = "Page_Info";
            LeftColumn = new SubmenuLeftColumn(0) { Parent = this };
            CenterColumn = new SubmenuCentralColumn(1) { Parent = this };
        }
        public void SwitchColumn(int index)
        {
            SwitchColumn((PM_COLUMNS)index);
        }

        public void SwitchColumn(PM_COLUMNS index)
        {
            switchColumnInternal(index, true);
        }

        private void switchColumnInternal(PM_COLUMNS index, bool apply)
        {
            var col = GetColumnAtPosition(index);
            CurrentColumnIndex = (int)index;
            if (Parent != null && Parent.Visible)
            {
                Main.PauseMenu._pause.CallFunction("SET_MENU_LEVEL", CurrentColumnIndex + 1);
                Main.PauseMenu._pause.CallFunction("MENU_SHIFT_DEPTH", 0, true, true);
                Parent.focusLevel = CurrentColumnIndex + 1;
                Main.PauseMenu._pause.CallFunction("SET_COLUMN_HIGHLIGHT", (int)col.position, col.Index, true, true);
                col.Items[col.Index].Selected = true;
            }
        }
        public void AddLeftItem(TabLeftItem item)
        {
            item.ParentTab = this;
            LeftColumn.AddItem(item);
        }

        public override void StateChange(int state)
        {
            Parent._pause._pause.CallFunction("MENU_STATE", (int)currentItemType);
            ((SubmenuCentralColumn)CenterColumn).Items.Clear();
            if(state != 0)
                ((SubmenuCentralColumn)CenterColumn).Items.AddRange(((TabLeftItem)LeftColumn.Items[LeftColumn.Index]).ItemList);
            ((SubmenuCentralColumn)CenterColumn).Items.ForEach(x => x.ParentColumn = ((SubmenuCentralColumn)CenterColumn));
            switch (currentItemType)
            {
                case LeftItemType.Statistics:
                    ((SubmenuCentralColumn)CenterColumn).VisibleItems = 16;
                    ((SubmenuCentralColumn)CenterColumn).InitColumnScroll(true, 2, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER);
                    ((SubmenuCentralColumn)CenterColumn).SetColumnScroll(-1, -1, -1, string.Empty, ((SubmenuCentralColumn)CenterColumn).Items.Count < ((SubmenuCentralColumn)CenterColumn).VisibleItems);
                    break;
                case LeftItemType.Settings:
                    ((SubmenuCentralColumn)CenterColumn).VisibleItems = 16;
                    ((SubmenuCentralColumn)CenterColumn).InitColumnScroll(true, 2, ScrollType.ALL, ScrollArrowsPosition.RIGHT);
                    ((SubmenuCentralColumn)CenterColumn).SetColumnScroll(((SubmenuCentralColumn)CenterColumn).Index + 1, ((SubmenuCentralColumn)CenterColumn).Items.Count, ((SubmenuCentralColumn)CenterColumn).VisibleItems, string.Empty, ((SubmenuCentralColumn)CenterColumn).Items.Count < ((SubmenuCentralColumn)CenterColumn).VisibleItems);
                    break;
                case LeftItemType.Info:
                    ((SubmenuCentralColumn)CenterColumn).VisibleItems = 10;
                    ((SubmenuCentralColumn)CenterColumn).InitColumnScroll(true, 2, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER);
                    ((SubmenuCentralColumn)CenterColumn).SetColumnScroll(-1, -1, -1, string.Empty, ((SubmenuCentralColumn)CenterColumn).Items.Count < ((SubmenuCentralColumn)CenterColumn).VisibleItems);
                    break;
                case LeftItemType.Keymap:
                    ((SubmenuCentralColumn)CenterColumn).VisibleItems = 15;
                    ((SubmenuCentralColumn)CenterColumn).InitColumnScroll(true, 2, ScrollType.UP_DOWN, ScrollArrowsPosition.CENTER);
                    ((SubmenuCentralColumn)CenterColumn).SetColumnScroll(-1, -1, -1, string.Empty, ((SubmenuCentralColumn)CenterColumn).Items.Count < ((SubmenuCentralColumn)CenterColumn).VisibleItems);
                    break;
                default:
                    ((SubmenuCentralColumn)CenterColumn).VisibleItems = 0;
                    break;
            }
        }

        public override void GoUp()
        {
            if (!Focused) return;
            switch (CurrentColumnIndex)
            {
                case 0:
                    LeftColumn.GoUp();
                    ((SubmenuCentralColumn)CenterColumn).currentColumnType = currentItemType;
                    StateChange((int)currentItemType);
                    Refresh(false);
                    Parent._pause._pause.CallFunction("SET_COLUMN_INPUT_EVENT", CurrentColumnIndex, 8);
                    break;
                case 1:
                    ((SubmenuCentralColumn)CenterColumn).GoUp();
                    if(((SubmenuCentralColumn)CenterColumn).currentColumnType == LeftItemType.Settings)
                        ((SubmenuCentralColumn)CenterColumn).SetColumnScroll(((SubmenuCentralColumn)CenterColumn).Index + 1, ((SubmenuCentralColumn)CenterColumn).Items.Count, ((SubmenuCentralColumn)CenterColumn).VisibleItems, string.Empty, ((SubmenuCentralColumn)CenterColumn).Items.Count < ((SubmenuCentralColumn)CenterColumn).VisibleItems);
                    break;
            }
        }

        public override void GoDown()
        {
            if (!Focused) return;
            switch (CurrentColumnIndex)
            {
                case 0:
                    LeftColumn.GoDown();
                    ((SubmenuCentralColumn)CenterColumn).currentColumnType = currentItemType;
                    StateChange((int)currentItemType);
                    Refresh(false);
                    Parent._pause._pause.CallFunction("SET_COLUMN_INPUT_EVENT", CurrentColumnIndex, 9);
                    break;
                case 1:
                    ((SubmenuCentralColumn)CenterColumn).GoDown();
                    if (((SubmenuCentralColumn)CenterColumn).currentColumnType == LeftItemType.Settings)
                        ((SubmenuCentralColumn)CenterColumn).SetColumnScroll(((SubmenuCentralColumn)CenterColumn).Index + 1, ((SubmenuCentralColumn)CenterColumn).Items.Count, ((SubmenuCentralColumn)CenterColumn).VisibleItems, string.Empty, ((SubmenuCentralColumn)CenterColumn).Items.Count < ((SubmenuCentralColumn)CenterColumn).VisibleItems);
                    break;
            }
        }

        public override async void MouseEvent(int eventType, int context, int index)
        {
            if (!Focused) return;
            switch (eventType)
            {
                case 5:
                    if (CurrentColumnIndex == context)
                    {
                        if (CurrentColumn.Index != index)
                        {
                            switch (CurrentColumnIndex)
                            {
                                case 0:
                                    LeftColumn.Items[LeftColumn.Index].Selected = false;
                                    LeftColumn.Index = index;
                                    LeftColumn.Items[LeftColumn.Index].Selected = true;
                                    StateChange((int)currentItemType);
                                    Refresh(false);
                                    break;
                                case 1:
                                    ((SubmenuCentralColumn)CenterColumn).Items[((SubmenuCentralColumn)CenterColumn).Index].Selected = false;
                                    ((SubmenuCentralColumn)CenterColumn).Index = index;
                                    ((SubmenuCentralColumn)CenterColumn).Items[((SubmenuCentralColumn)CenterColumn).Index].Selected = true;
                                    break;
                            }
                            return;
                        }
                        switch (CurrentColumnIndex)
                        {
                            case 0 when currentItemType == LeftItemType.Settings:
                                TabLeftItem leftItem = (TabLeftItem)LeftColumn.Items[LeftColumn.Index];
                                if (!leftItem.Enabled)
                                {
                                    Game.PlaySound(TabView.AUDIO_ERROR, TabView.AUDIO_LIBRARY);
                                    return;
                                }
                                CurrentColumnIndex++;
                                if (leftItem.ItemList.All(x => !(x as SettingsItem).Enabled)) break;
                                while (!(((SubmenuCentralColumn)CenterColumn).Items[((SubmenuCentralColumn)CenterColumn).Index] as SettingsItem).Enabled)
                                {
                                    await BaseScript.Delay(0);
                                    ((SubmenuCentralColumn)CenterColumn).Index++;
                                }
                                Parent.FocusLevel++;
                                Parent.SendColumnItemSelect(LeftColumn);
                                break;
                            case 1:
                                ((SubmenuCentralColumn)CenterColumn).Select();
                                Parent.SendColumnItemSelect(((SubmenuCentralColumn)CenterColumn));
                                break;
                        }
                    }
                    else
                    {
                        if (context > CurrentColumnIndex)
                        {
                            Parent.FocusLevel++;
                            CurrentColumnIndex++;
                        }
                        else if (context < CurrentColumnIndex)
                        {
                            Parent.FocusLevel--;
                            CurrentColumnIndex--;
                        }
                        switch (CurrentColumnIndex)
                        {
                            case 0:
                                LeftColumn.Items[LeftColumn.Index].Selected = false;
                                LeftColumn.Index = index;
                                LeftColumn.Items[LeftColumn.Index].Selected = true;
                                StateChange((int)currentItemType);
                                Refresh(false);
                                break;
                            case 1:
                                ((SubmenuCentralColumn)CenterColumn).Items[((SubmenuCentralColumn)CenterColumn).Index].Selected = false;
                                ((SubmenuCentralColumn)CenterColumn).Index = index;
                                ((SubmenuCentralColumn)CenterColumn).Items[((SubmenuCentralColumn)CenterColumn).Index].Selected = true;
                                break;
                        }
                    }
                    break;
                case 10: // mouse wheel up
                case 11: // mouse wheel down
                    MouseScroll(eventType == 10 ? -1 : 1);
                    break;
            }
        }

        public void MouseScroll(int dir)
        {
            int col = Parent.hoveredColumn;
            switch (CurrentColumnIndex)
            {
                case 0:
                    if (col == 1)
                    {
                        if (currentItemType == LeftItemType.Info || currentItemType == LeftItemType.Statistics)
                        {
                            Debug.WriteLine("GET_HOVERED_COLUMN -- Current: " + col);
                            Game.PlaySound(TabView.AUDIO_UPDOWN, TabView.AUDIO_LIBRARY);
                            return;
                        }
                    }
                    if (dir == -1)
                        LeftColumn.GoUp();
                    else
                        LeftColumn.GoDown();
                    ((SubmenuCentralColumn)CenterColumn).currentColumnType = currentItemType;
                    StateChange((int)currentItemType);
                    Refresh(false);
                    break;
                case 1:
                    if (currentItemType == LeftItemType.Settings)
                    {
                        if (dir == -1)
                            ((SubmenuCentralColumn)CenterColumn).GoUp();
                        else
                            ((SubmenuCentralColumn)CenterColumn).GoDown();
                    }
                    if (((SubmenuCentralColumn)CenterColumn).currentColumnType == LeftItemType.Settings)
                        ((SubmenuCentralColumn)CenterColumn).SetColumnScroll(((SubmenuCentralColumn)CenterColumn).Index + 1, ((SubmenuCentralColumn)CenterColumn).Items.Count, ((SubmenuCentralColumn)CenterColumn).VisibleItems, string.Empty, ((SubmenuCentralColumn)CenterColumn).Items.Count < ((SubmenuCentralColumn)CenterColumn).VisibleItems);
                    Game.PlaySound(TabView.AUDIO_UPDOWN, TabView.AUDIO_LIBRARY);
                    break;
            }
        }

        public override void GoLeft()
        {
            if(!Focused) return;
            Parent._pause._pause.CallFunction("SET_COLUMN_INPUT_EVENT", CurrentColumnIndex, 10);

            if (CurrentColumnIndex == 1)
            {
                ((SubmenuCentralColumn)CenterColumn).GoLeft();
            }
        }

        public override void GoRight()
        {
            if (!Focused) return;
            Parent._pause._pause.CallFunction("SET_COLUMN_INPUT_EVENT", CurrentColumnIndex, 11);
            if (CurrentColumnIndex == 1)
            {
                ((SubmenuCentralColumn)CenterColumn).GoRight();
            }
        }

        public override async void Select()
        {
            if (!Focused) return;
            switch (CurrentColumnIndex)
            {
                case 0:
                    if (currentItemType != LeftItemType.Settings) return;
                    TabLeftItem leftItem = (TabLeftItem)LeftColumn.Items[LeftColumn.Index];
                    if (!leftItem.Enabled)
                    {
                        Game.PlaySound(TabView.AUDIO_ERROR, TabView.AUDIO_LIBRARY);
                        return;
                    }
                    CurrentColumnIndex++;
                    if (leftItem.ItemList.All(x => !(x as SettingsItem).Enabled)) break;
                    while (!(((SubmenuCentralColumn)CenterColumn).Items[((SubmenuCentralColumn)CenterColumn).Index] as SettingsItem).Enabled)
                    {
                        await BaseScript.Delay(0);
                        ((SubmenuCentralColumn)CenterColumn).Index++;
                    }
                    Parent.FocusLevel++;
                    Parent.SendColumnItemSelect(LeftColumn);
                    break;
                case 1:
                    ((SubmenuCentralColumn)CenterColumn).Select();
                    Parent.SendColumnItemSelect(((SubmenuCentralColumn)CenterColumn));
                    break;
            }
        }

        public override void GoBack()
        {
            if (!Focused) return;
            if (CurrentColumnIndex == 1)
            {
                CurrentColumnIndex--;
                Parent.FocusLevel--;
            }
        }

        public override void Focus()
        {
            base.Focus();
            LeftColumn.Index = LeftColumn.index;
            LeftColumn.HighlightColumn(true, false, true);
            LeftColumn.Items[LeftColumn.Index].Selected = true;
            Refresh(true);
        }

        public override void UnFocus()
        {
            base.UnFocus();
            LeftColumn.Items[LeftColumn.Index].Selected = false;
        }

        public override void Refresh(bool highlightOldIndex)
        {
            Parent._pause._pause.CallFunction("ALLOW_CLICK_FROM_COLUMN", 0, true);
            Parent._pause._pause.CallFunction("SET_DATA_SLOT_EMPTY", 1);
            for (int i = 0; i < ((SubmenuCentralColumn)CenterColumn).Items.Count; i++)
            {
                SetDataSlot(((SubmenuCentralColumn)CenterColumn).position, i);
            }
            switch (currentItemType)
            {
                case LeftItemType.Keymap:
                    Parent._pause._pause.CallFunction("SET_COLUMN_TITLE", 1, ((TabLeftItem)LeftColumn.Items[LeftColumn.Index]).RightTitle, ((TabLeftItem)LeftColumn.Items[LeftColumn.Index]).KeymapRightLabel_1, ((TabLeftItem)LeftColumn.Items[LeftColumn.Index]).KeymapRightLabel_2);
                    Parent._pause._pause.CallFunction("SET_COLUMN_FOCUS", 1, false, false, false);
                    break;
                case LeftItemType.Settings:
                    if (highlightOldIndex)
                        Parent._pause._pause.CallFunction("SET_COLUMN_HIGHLIGHT", 1, ((SubmenuCentralColumn)CenterColumn).Index, true, true);
                        //Parent._pause._pause.CallFunction("SET_COLUMN_STATE", 3);
                        ////Parent._pause._pause.CallFunction("SET_CONTROL_IMAGE", "pause_menu_pages_settings_pc", "controller");
                        ////Parent._pause._pause.CallFunction("SET_CONTROL_LABELS", "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "SUPER TITLE");

                        ////Parent._pause._pause.CallFunction("SET_VIDEO_MEMORY_BAR", true, "ScaleformUI Pause Menu Awesomeness", 70, 116);
                        //Parent._pause._pause.CallFunction("SET_DESCRIPTION", 1, "~r~Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat", "scaleformui", "pauseinfobg", 0, 0, 578, 160);
                        break;
            }
            ((SubmenuCentralColumn)CenterColumn).ShowColumn();
        }

        public override void Populate()
        {
            TabLeftItem item = (TabLeftItem)LeftColumn.Items[LeftColumn.Index];
            item.Selected = true;
            ((SubmenuCentralColumn)CenterColumn).Items.Clear();
            if (currentItemType != LeftItemType.Empty)
            {
                ((SubmenuCentralColumn)CenterColumn).Items.AddRange(item.ItemList);
                ((SubmenuCentralColumn)CenterColumn).Items.ForEach(x => x.ParentColumn = ((SubmenuCentralColumn)CenterColumn));
            }
            Parent._pause._pause.CallFunction("SET_MENU_LEVEL", 1);
            Parent._pause._pause.CallFunction("MENU_STATE", (int)currentItemType);
            Parent._pause._pause.CallFunction("SET_MENU_LEVEL", 0);
            for (int i = 0; i < LeftColumn.Items.Count; i++)
            {
                SetDataSlot(LeftColumn.position, i);
            }
            for (int i = 0; i < ((SubmenuCentralColumn)CenterColumn).Items.Count; i++)
            {
                SetDataSlot(((SubmenuCentralColumn)CenterColumn).position, i);
            }
        }

        public override void ShowColumns()
        {
            //Parent._pause._pause.CallFunction("MOUSE_COLUMN_SHIFT", 0);
            //Parent._pause._pause.CallFunction("SET_COLUMN_HIGHLIGHT", 0, 0, false, false);
            LeftColumn.ShowColumn();
            ((SubmenuCentralColumn)CenterColumn).ShowColumn();
            if (currentItemType == LeftItemType.Settings)
            {
                Parent._pause._pause.CallFunction("SET_COLUMN_STATE", 0);
            }
            Parent._pause._pause.CallFunction("SET_COLUMN_FOCUS", 0, false, false, false);
            //to be overridden by subcolumns
            LeftColumn.InitColumnScroll(true, 1, ScrollType.UP_DOWN, ScrollArrowsPosition.RIGHT);
            LeftColumn.SetColumnScroll(LeftColumn.Index, LeftColumn.Items.Count, 16, "", LeftColumn.Items.Count < 16);
        }

        public override void SetDataSlot(PM_COLUMNS slot, int index)
        {
            switch (slot)
            {
                case PM_COLUMNS.LEFT:
                    LeftColumn.SetDataSlot(index);
                    break;
                case PM_COLUMNS.MIDDLE:
                    ((SubmenuCentralColumn)CenterColumn).SetDataSlot(index);
                    break;
            }

        }

        public override void UpdateSlot(PM_COLUMNS slot, int index)
        {
            switch (slot)
            {
                case PM_COLUMNS.LEFT:
                    LeftColumn.UpdateSlot(index);
                    break;
                case PM_COLUMNS.MIDDLE:
                    ((SubmenuCentralColumn)CenterColumn).UpdateSlot(index);
                    break;
            }
        }
    }
}